### PR TITLE
[Feature] Add DMCP and fix the deploy pipeline of NAS algorithms

### PR DESCRIPTION
**Pull request by Lxtccc**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here and update the documentation.

## Checklist

**Before PR**:

- [ ] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmrazor/models/algorithms/pruning/dmcp.py [hunk 1]
```diff
@@ -0,0 +1,403 @@
...
+
+            # According to fix_subnet, delete the unchosen part of supernet
+            load_fix_subnet(self.architecture, fix_subnet)
+            self.is_supernet = False
+        self.is_deployed = is_deployed
+        if (self.is_deployed):
+            # To static ops for loaded pruned network.
+            self._deploy()
+    
+    def _deploy(self):
```
Comment (sunnyxiaohu) [inline mmrazor/models/algorithms/pruning/dmcp.py:89]: `_deploy` is not needed here.

File: mmrazor/models/algorithms/pruning/dmcp.py [hunk 2]
```diff
@@ -0,0 +1,403 @@
...
+from mmrazor.models.mutators import ChannelMutator
+from .ite_prune_algorithm import ItePruneAlgorithm, ItePruneConfigManager
+
+LossResults = Dict[str, torch.Tensor]
+TensorResults = Union[Tuple[torch.Tensor], torch.Tensor]
+PredictResults = List[BaseDataElement]
+ForwardResults = Union[LossResults, TensorResults, PredictResults]
+
+@MODELS.register_module()
+class DMCP(ItePruneAlgorithm):
```
Comment (sunnyxiaohu) [inline mmrazor/models/algorithms/pruning/dmcp.py:35]: inherit from ItePruneAlgorithm for ?

File: mmrazor/models/mutators/channel_mutator/dmcp_channel_mutator.py [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+            Defaults to dict( type='BackwardTracer',
+                loss_calculator=dict(type='ImageClassifierPseudoLoss')).
+            Change loss_calculator according to task and backbone.
+        pruning_cfg (Tuple): (min_sample_rate, max_sample_rate, sample_offset)).
+    """
+
+    def __init__(self,
+                 channel_unit_cfg: Union[dict, Type[ChannelUnitType]] = dict(
+                     type='DMCPChannelUnit', units={}),
+                 parse_cfg=dict(
```
Comment (sunnyxiaohu) [inline mmrazor/models/mutators/channel_mutator/dmcp_channel_mutator.py:32]: update parse_cfg to `ChannelAnalyzer`, according to: https://github.com/open-mmlab/mmrazor/blob/26da4d223a1c215388b1a911317e5921cc288917/mmrazor/models/mutators/channel_mutator/channel_mutator.py#L73

File: mmrazor/models/mutators/channel_mutator/dmcp_channel_mutator.py [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+                return direct_channel
+            elif mode == 'expected':
+                marginal_prob = torch.cumprod(condition_prob, dim=0)
+                expected_channel = (torch.sum(marginal_prob) *
+                                    group_size) + min_ch
+                return expected_channel
+            else:
+                raise NotImplementedError
+
+    def calc_current_flops(self, model):
```
Comment (sunnyxiaohu) [inline mmrazor/models/mutators/channel_mutator/dmcp_channel_mutator.py:153]: move to Algorithm.

File: mmrazor/models/task_modules/estimators/counters/flops_params_counter.py [hunk 5]
```diff
@@ -172,7 +171,8 @@ def params_units_convert(num_params, units='M', precision=3):
         >>> params_units_convert(3e-9)
         '3e-09'
     """
-
+    if getattr(num_params, 'requires_grad', None):
```
Comment (sunnyxiaohu) [inline mmrazor/models/task_modules/estimators/counters/flops_params_counter.py:174]: If you want to get num_params without units, just pass the argument `units` with the `None`.

**Review by sunnyxiaohu — APPROVED** [review]

File: mmrazor/models/architectures/dynamic_ops/bricks/dynamic_norm.py [hunk 6]
```diff
@@ -257,118 +257,70 @@ def _check_input_dim(self, input: Tensor) -> None:
                 input.dim()))
 
 
-class DynamicSyncBatchNorm(nn.SyncBatchNorm, DynamicBatchNormMixin):
-    """DynamicOp for sync bn."""
+@MODELS.register_module()
+class DMCPBatchNorm2d(DynamicBatchNorm2d):
```
Comment (sunnyxiaohu) [inline mmrazor/models/architectures/dynamic_ops/bricks/dynamic_norm.py:261]: fix `dynamic_norm.py`

File: mmrazor/models/mutators/channel_mutator/dmcp_channel_mutator.py [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+
+    def sample_subnet(self, mode, arch_train):
+        choices = dict()
+        for group_id, _ in self.search_groups.items():
+            choices[group_id] = self._prune_by_arch(mode, group_id)
+        self.set_choices(choices)
+
+        self.modify_supernet_forward(arch_train)
+
+    def _prune_by_arch(self, mode, group_id):
```
Comment (Lxtccc) [inline mmrazor/models/mutators/channel_mutator/dmcp_channel_mutator.py:116]: docstr

File: mmrazor/models/mutators/channel_mutator/dmcp_channel_mutator.py [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+        self.arch_params = nn.ParameterDict()
+        self._op_arch_align = dict()
+        self._arch_params_attr = dict()
+        for group_id, module in self.search_groups.items():
+            arch_message = self._generate_arch_message(module[0].mutable_channel.num_channels)
+            self._arch_params_attr[str(group_id)] = arch_message
+            group_arch_param = self._build_arch_param(arch_message[1])
+            self.arch_params[str(group_id)] = group_arch_param
+
+            for m in module[0].output_related:
```
Comment (Lxtccc) [inline mmrazor/models/mutators/channel_mutator/dmcp_channel_mutator.py:66]: modify name:m->unit

File: mmrazor/models/mutables/mutable_channel/units/dmcp_channel_unit.py [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+
+    def __init__(self,
+                 num_channels: int,
+                 choice_mode: str = 'number',
+                 divisor: int = 1,
+                 min_value: int = 1,
+                 min_ratio: float = 0.5) -> None:
+        super().__init__(num_channels, choice_mode, divisor, min_value,
+                         min_ratio)
+        self._traceable_choice = tracable_choice()
```
Comment (Lxtccc) [inline mmrazor/models/mutables/mutable_channel/units/dmcp_channel_unit.py:47]: traceable

File: mmrazor/models/architectures/dynamic_ops/bricks/dynamic_norm.py [hunk 10]
```diff
@@ -388,3 +391,101 @@ def static_op_factory(self):
...
+    accepted_mutable_attrs = {'num_features'}
+
+    def __init__(self, *args, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
+        self.mutable_attrs: Dict[str, Optional[BaseMutable]] = nn.ModuleDict()
+
+    @classmethod
+    def convert_from(cls, module: _BatchNorm):
```
Comment (sunnyxiaohu) [inline mmrazor/models/architectures/dynamic_ops/bricks/dynamic_norm.py:405]: del

**Review by sunnyxiaohu — COMMENTED** [review]
1. save the export subnet info by "units" (ref: DCFF)
2. I suggest to remove all the retraining logic in Algorithm and use `sub_model` instead.

File: configs/pruning/mmcls/dcff/fix_subnet.json [hunk 11]
```diff
@@ -14,6 +14,102 @@
                     "min_value":1,
                     "min_ratio":0.9
                 },
+                "channels":{
```
Comment (sunnyxiaohu) [inline configs/pruning/mmcls/dcff/fix_subnet.json:17]: Could the "channels" field be hidden when exporting and loading subnet?

File: mmrazor/models/algorithms/nas/dsnas.py [hunk 12]
```diff
@@ -75,24 +71,18 @@ def __init__(self,
...
+        # provides some APIs commonly used by NAS.
+        # Before using it, you must do some preparation according to
+        # the supernet.
+        self.mutator.prepare_from_supernet(self.architecture)
+        self.mutator.prepare_arch_params()
+
+        self.mutable_module_resources = self._get_module_resources()
+        self.search_space_name_list = list(self.mutator._name2mutable.keys())
+
+        self.is_supernet = True
```
Comment (sunnyxiaohu) [inline mmrazor/models/algorithms/nas/dsnas.py:85]: field "is_supernet" is not required.

File: tests/test_models/test_algorithms/test_dmcp.py [hunk 13]
```diff
@@ -137,12 +133,6 @@ def test_dmcp_train_step(self):
         assert losses['random_subnet1.loss'] > 0
         assert losses['random_subnet2.loss'] > 0
 
-    def test_dmcp_load_fix_subnet(self):
```
Comment (sunnyxiaohu) [inline tests/test_models/test_algorithms/test_dmcp.py]: Fix this UT not just delete it.

File: tests/test_models/test_algorithms/test_darts.py [hunk 14]
```diff
@@ -129,16 +123,6 @@ def test_forward_loss(self) -> None:
         loss = algo(inputs, mode='loss')
         self.assertIsInstance(loss, dict)
 
-        # subnet
```
Comment (sunnyxiaohu) [inline tests/test_models/test_algorithms/test_darts.py:132]: Add UT: `test_load_subnet`

File: tests/test_models/test_algorithms/test_dsnas.py [hunk 15]
```diff
@@ -116,12 +111,6 @@ def test_forward_loss(self) -> None:
         loss = algo(inputs, mode='loss')
         self.assertIsInstance(loss, dict)
 
-        # subnet
```
Comment (sunnyxiaohu) [inline tests/test_models/test_algorithms/test_dsnas.py:119]: Add UT: test_load_subnet

File: tests/test_models/test_algorithms/test_spos.py [hunk 16]
```diff
@@ -57,11 +57,6 @@ def test_init(self):
         alg = SPOS(model, mutator)
         self.assertIsInstance(alg.mutator, NasMutator)
 
-        # initiate spos when `fix_subnet` is not None.
```
Comment (sunnyxiaohu) [inline tests/test_models/test_algorithms/test_spos.py:60]: Add UT: test_load_subnet

File: configs/nas/mmcls/darts/darts_subnet_1xb96_cifar10_2.0.py [hunk 17]
```diff
@@ -29,13 +29,11 @@
...
-    type='mmrazor.SPOS',
-    architecture=supernet,
-    mutator=None,
-    fix_subnet=fix_subnet,
-)
+    _scope_='mmrazor',
+    type='sub_model',
+    cfg=supernet,
+    fix_subnet='configs/nas/mmcls/darts/DARTS_SUBNET_CIFAR_PAPER_ALIAS.yaml',
+    mode='mutator')
```
Comment (sunnyxiaohu) [inline configs/nas/mmcls/darts/darts_subnet_1xb96_cifar10_2.0.py:37]: mode='mutable'

File: configs/nas/mmcls/darts/darts_subnet_1xb96_cifar10_2.0_mmrazor.py [hunk 18]
```diff
@@ -29,13 +29,11 @@
...
-    type='mmrazor.SPOS',
-    architecture=supernet,
-    mutator=None,
-    fix_subnet=fix_subnet,
-)
+    _scope_='mmrazor',
+    type='sub_model',
+    cfg=supernet,
+    fix_subnet='configs/nas/mmcls/darts/DARTS_SUBNET_CIFAR_MMRAZOR_97.32.yaml',
+    mode='mutator')
```
Comment (sunnyxiaohu) [inline configs/nas/mmcls/darts/darts_subnet_1xb96_cifar10_2.0_mmrazor.py:37]: mode='mutable'

File: configs/nas/mmcls/dsnas/dsnas_subnet_8xb128_in1k.py [hunk 19]
```diff
@@ -1,8 +1,16 @@
...
+supernet = _base_.model['architecture']
 
-model = dict(fix_subnet=fix_subnet)
+model_cfg = dict(
+    _scope_='mmrazor',
+    type='sub_model',
+    cfg=supernet,
+    fix_subnet=  # noqa: E251
+    'configs/nas/mmcls/dsnas/DSNAS_SUBNET_IMAGENET_PAPER_ALIAS.yaml',  # noqa: E501
+    mode='mutator')
```
Comment (sunnyxiaohu) [inline configs/nas/mmcls/dsnas/dsnas_subnet_8xb128_in1k.py:12]: mode='mutable'

File: configs/nas/mmcls/onceforall/ofa_mobilenet_subnet_8xb256_in1k.py [hunk 20]
```diff
@@ -1,6 +1,14 @@
...
-model = dict(
-    fix_subnet='configs/nas/mmcls/onceforall/OFA_SUBNET_NOTE8_LAT22.yaml')
+supernet = _base_.supernet
+
+model_cfg = dict(
+    _scope_='mmrazor',
+    type='sub_model',
+    cfg=supernet,
+    fix_subnet='configs/nas/mmcls/onceforall/OFA_SUBNET_NOTE8_LAT22.yaml',
+    mode='mutator')
```
Comment (sunnyxiaohu) [inline configs/nas/mmcls/onceforall/ofa_mobilenet_subnet_8xb256_in1k.py:10]: mode='mutable'

File: configs/nas/mmcls/spos/spos_mobilenet_subnet_8xb128_in1k.py [hunk 21]
```diff
@@ -1,8 +1,15 @@
...
-fix_subnet = 'configs/nas/spos/AngleNAS_SHUFFLENETV2_IN1k_2.0.yaml'  # noqa: E501
+supernet = _base_.supernet
 
-model = dict(fix_subnet=fix_subnet)
+model_cfg = dict(
+    _scope_='mmrazor',
+    type='sub_model',
+    cfg=supernet,
+    fix_subnet='configs/nas/spos/AngleNAS_SHUFFLENETV2_IN1k_2.0.yaml',
+    mode='mutator')
```
Comment (sunnyxiaohu) [inline configs/nas/mmcls/spos/spos_mobilenet_subnet_8xb128_in1k.py:11]: mode='mutable'

File: configs/nas/mmcls/spos/spos_shufflenet_subnet_8xb128_in1k.py [hunk 22]
```diff
@@ -1,8 +1,15 @@
...
-fix_subnet = 'configs/nas/mmcls/spos/SPOS_SUBNET.yaml'
+supernet = _base_.supernet
 
-model = dict(fix_subnet=fix_subnet)
+model_cfg = dict(
+    _scope_='mmrazor',
+    type='sub_model',
+    cfg=supernet,
+    fix_subnet='configs/nas/mmcls/spos/SPOS_SUBNET.yaml',
+    mode='mutator')
```
Comment (sunnyxiaohu) [inline configs/nas/mmcls/spos/spos_shufflenet_subnet_8xb128_in1k.py:11]: mode='mutable'

File: configs/nas/mmdet/detnas/detnas_shufflenet_subnet_8xb128_in1k.py [hunk 23]
```diff
@@ -1,8 +1,16 @@
...
+supernet = _base_.supernet
 
-model = dict(fix_subnet=fix_subnet)
+model_cfg = dict(
+    _scope_='mmrazor',
+    type='sub_model',
+    cfg=supernet,
+    fix_subnet=  # noqa: E251
+    'https://download.openmmlab.com/mmrazor/v1/detnas/detnas_subnet_frcnn_shufflenetv2_fpn_1x_coco_bbox_backbone_flops-0.34M_mAP-37.5_20220715-61d2e900_subnet_cfg_v1.yaml',  # noqa: E501
+    mode='mutator')
```
Comment (sunnyxiaohu) [inline configs/nas/mmdet/detnas/detnas_shufflenet_subnet_8xb128_in1k.py:12]: mode='mutable'

File: configs/nas/mmdet/detnas/detnas_frcnn_shufflenet_subnet_coco_1x.py [hunk 24]
```diff
@@ -1,8 +1,15 @@
...
-fix_subnet = 'configs/nas/mmdet/detnas/DETNAS_SUBNET.yaml'
+supernet = _base_.supernet
 
-model = dict(fix_subnet=fix_subnet)
+model_cfg = dict(
+    _scope_='mmrazor',
+    type='sub_model',
+    cfg=supernet,
+    fix_subnet='configs/nas/mmdet/detnas/DETNAS_SUBNET.yaml',
+    mode='mutator')
```
Comment (sunnyxiaohu) [inline configs/nas/mmdet/detnas/detnas_frcnn_shufflenet_subnet_coco_1x.py:11]: mode='mutable'

File: configs/pruning/mmcls/dcff/fix_subnet.json (hunk 11 above)
Comment (Lxtccc) [reply configs/pruning/mmcls/dcff/fix_subnet.json:17]: Subject to the way channel_unit is initialized, the pruned model can only be constructed at this stage if the channel is included at export and load time. At least for pruning algorithms.

File: configs/nas/mmcls/autoformer/README.md [hunk 25]
```diff
@@ -43,7 +43,7 @@ CUDA_VISIBLE_DEVICES=0,1,2,3 PORT=29500 ./tools/dist_train.sh \
 
 ```bash
 CUDA_VISIBLE_DEVICES=0 PORT=29500 ./tools/dist_test.sh \
-  configs/nas/mmcls/autoformer/autoformer_subnet_8xb128_in1k.py \
+  configs/nas/mmcls/autoformer/autoformer_supernet_8xb128_in1k.py \
```
[the remark this answers is not in the document]
Comment (Lxtccc) [reply configs/nas/mmcls/autoformer/README.md:46]: A new autoformer_subnet_8xb128_in1k.py config will be created for the third stage

File: mmrazor/models/algorithms/nas/dsnas.py (hunk 12 above)
Comment (Lxtccc) [reply mmrazor/models/algorithms/nas/dsnas.py:85]: In dsnas, ‘is_supernet’ is used in lines 133/138/288/193 for the initialization of the finetune phase.

File: configs/nas/mmcls/bignas/ATTENTIVE_SUBNET_A0.yaml [hunk 26]
```diff
@@ -1,64 +1,72 @@
 backbone.first_channels:
-  chosen: 16
+  chosen: 24
```
Comment (aptsunny) [inline configs/nas/mmcls/bignas/ATTENTIVE_SUBNET_A0.yaml:2]: 这里显然不应该动，24就不是A0的配置了

File: mmrazor/registry/registry.py [hunk 27]
```diff
@@ -113,7 +113,8 @@ def sub_model(cfg,
               prefix: str = '',
               extra_prefix: str = '',
               init_weight_from_supernet: bool = False,
-              init_cfg: Optional[Dict] = None):
+              init_cfg: Optional[Dict] = None,
+              **kwargs):
```
Comment (gaoyang07) [inline mmrazor/registry/registry.py:117]: `**kwargs` is not used in this method.

File: configs/nas/mmcls/onceforall/ofa_mobilenet_subnet_8xb256_in1k.py [hunk 28]
```diff
@@ -1,6 +1,18 @@
...
+_base_.supernet.data_preprocessor = _base_.data_preprocessor
+
+supernet = _base_.supernet
+
+model_cfg = dict(
+    _scope_='mmrazor',
+    type='sub_model',
+    cfg=supernet,
+    fix_subnet='configs/nas/mmcls/onceforall/OFA_SUBNET_NOTE8_LAT31.yaml')
+# fix_subnet='configs/nas/mmcls/onceforall/OFA_SUBNET_NOTE8_LAT22.yaml')
```
Comment (gaoyang07) [inline configs/nas/mmcls/onceforall/ofa_mobilenet_subnet_8xb256_in1k.py:12]: remove this line.

File: mmrazor/structures/subnet/fix_subnet.py [hunk 29]
```diff
@@ -162,7 +170,15 @@ def export_fix_subnet(
 
         if next(copied_model.parameters()).is_cuda:
             copied_model.cuda()
-        return fix_subnet, copied_model
+
+        if remove_architecture:
```
Comment (gaoyang07) [inline mmrazor/structures/subnet/fix_subnet.py:174]: solve `architecture` key error by adding prefix when initializing weights

File: mmrazor/engine/runner/evolution_search_loop.py [hunk 30]
```diff
@@ -320,10 +320,7 @@ def _save_best_fix_subnet(self):
             timestamp_subnet = time.strftime('%Y%m%d_%H%M', time.localtime())
             model_name = f'subnet_{timestamp_subnet}.pth'
             save_path = osp.join(self.runner.work_dir, model_name)
-            torch.save({
-                'state_dict': sliced_model.state_dict(),
```
Comment (gaoyang07) [inline mmrazor/engine/runner/evolution_search_loop.py:324]: rollback maybe.

File: configs/nas/mmcls/autoformer/autoformer_subnet_8xb256_in1k.py [hunk 31]
```diff
@@ -0,0 +1,13 @@
+_base_ = 'autoformer_supernet_32xb256_in1k.py'
+
+supernet = _base_.supernet
+
+model_cfg = dict(
+    _scope_='mmrazor',
+    type='sub_model',
+    cfg=supernet,
+    fix_subnet='STEP2_SUBNET_YAML.yaml')
```
Comment (gaoyang07) [inline configs/nas/mmcls/autoformer/autoformer_subnet_8xb256_in1k.py:9]: Add subnet config yaml for searched AutoFormer.

**Review by aptsunny — APPROVED** [review]

**Comment by vansin** [comment]
@Lxtccc Congratulations on having your first PR merged in OpenMMLab! I am a technical operations staff member of OpenMMLab. My WeChat ID is **van-sin** ,please add me . I would like to invite you to join the OpenMMLab contributors WeChat group.

**Pull request merged**